### ISSUE TITLE
Issue with "Object.prototype.hasOwnProperty()": polyfill is incorrect

### ISSUE THREAD

**Issue by cpcallen** [mdn/content#3626]
MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty#polyfill

#### What information was incorrect, unhelpful, or incomplete?

The polyfill is not semantically equivalent to the algorithm given in the specification.  Notably, given

```JS
var p = {p: 1}, o = Object.create(p);
o.p = 1;
```

`o.hasOwnProperty('p')` returns true, but the polyfill returns false.

#### Specific section or headline?

Polyfill

#### What did you expect to see?

It would probably be most useful if the given polyfill installed the following method instead:

```JS
  function hasOwnProperty(key) {
    if (this == null) { throw new TypeError("Cannot convert undefined or null to object!") }
    var o = Object(this);
    return Object.getOwnPropertyDescriptor(o, key) !== null;
  }
```

It might be argued that this polyfill is not as practically useful as the one currently given, in that `Object.getOwnPropertyDescriptor` was introduced in ES5 while `Object.prototype.hasOwnProperty` dates to ES3—but I would counter that:

* `Object.prototype.__proto__` was not standardised until ES6 (the better alternative, `Object.getPrototypeOf`, until ES5).
* Because `hasOwnProperty` has been around since ES3 it is very unlikely that anyone needs a polyfill for practical purposes at this point.  The polyfill section of the documentation is instead more useful as a more rigorous statement of the algorithm implemented, and as such it ought to prioritise correctness over other concerns.

(I am not claiming that my proposed polyfill is 100% correct, but I do claim that it is strictly more correct than the current one.)

#### Did you test this? If so, how?

Careful examination of the ES3, ES5 and ES2020 specifications as well as testing the existing and suggested polyfills manually in in node.js v12.21.0.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/object/hasownproperty`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/object/hasownproperty/index.html
* Last commit: https://github.com/mdn/content/commit/3b80812a1d4f68e57f35fca763f50bf3e5dfe997
* Document last modified: 2021-02-19T19:57:02.000Z

</details>

**Comment by evilpie** [comment]
I don't know if there is some policy for polyfills, but this one doesn't seem that useful. The function was added a long time ago.

**Comment by sideshowbarker** [comment]
> I don't know if there is some policy for polyfills, but this one doesn't seem that useful. The function was added a long time ago.

Yeah, so the best fix here might be to just drop that entire Polyfill from the article. @cpcallen would that be OK with you? Or do you think it’s still useful to have a (correct) polyfill in the article?

**Comment by sideshowbarker** [comment]
I went ahead and opened https://github.com/mdn/content/pull/3658 with a patch that drops the whole Polyfill section. But if @cpcallen prefers it be fixed rather than cut, we could close that PR and open another one.

**Comment by cpcallen** [comment]
As a JavaScript interpreter author, I have found polyfills on MDN to be very useful for two purposes:

* They often provide a very clear explanation of the mechanism implemented by a particular function that is more exact than the textual description but easier to understand than the specification.  I may need to refer to the spec to get all the details, but the polyfill makes the spec vastly more comprehensible.

* In most cases, I can use the polyfill as-is either permanently (if performance is not an issue) or temporarily until I write a native implementation.

With that in mind, I'd say that no polyfill is preferable to a misleading one, but a correct one would be better.  You are welcome to use the one I provided above, though I would invite you to check it for errors first.

**Comment by sideshowbarker** [comment]
> As a JavaScript interpreter author, I have found polyfills on MDN to be very useful for two purposes:
> 
> * They often provide a very clear explanation of the mechanism implemented by a particular function that is more exact than the textual description but easier to understand than the specification.  I may need to refer to the spec to get all the details, but the polyfill makes the spec vastly more comprehensible.
> * In most cases, I can use the polyfill as-is either permanently (if performance is not an issue) or temporarily until I write a native implementation.

Those are clearly reasonable purposes but in practice at MDN the primary purpose of including polyfills is for web developers who need them in cases where there are a significant number of their users running browsers which don’t support something natively.

So unless there’s something exceptional about a particular feature that merits making a polyfill for it available at MDN even after the feature has been implemented in deployed browsers for years, it seems best not to keep maintaining the polyfill for it at MDN.

And since that seems to be he case here, I’m going ahead an merge the PR for dropping it.

> With that in mind, I'd say that no polyfill is preferable to a misleading one, but a correct one would be better.

100% agreed

> You are welcome to use the one I provided above, though I would invite you to check it for errors first.

If it were the case that the polyfill was one that web developers in general were getting some utility from in order get their code working for users in older browsers they need to target, I would be willing to spend time checking the polyfill against the spec requirements. But since that doesn’t seem to be the case here, it seems better to drop the polyfill from MDN.

**Comment by cpcallen** [comment]
No complaints about how this issue in particular has been addressed, but I wanted to address @sideshowbarker's comment a little bit.

I see a pattern in the response to most of the bugs I file against MDN, where the response is something along the lines of "your use case is not important  to us, because most of our users are web developers and so don't benefit from <whatever it was that I requested>".

Explanatory polyfills are one example; another are the links that MDN formerly had to the version of the spec in which each feature was introduced.  (It now just links to the most recent, actually not-yet-formally-published version fo the spec.  The argument against reverting this change was that the previous arrangement was regarded as harmful because people were being directed to 'out of date' information.  I concede the point but there were ways to solve this problem without removing useful version information.)

It's hard to argue against any individual change of this kind, but the "this feature is not useful to our core audience" attitude becomes self-fulfilling, because MDN has become increasingly less useful to its many other users.  When I started working on JavaScript implementations in 2017 I would always begin by looking at MDN and could usually do an initial implementation of a feature based on the material here; I would eventually have to go read the ECMAScript spec to get the corner cases correct, but I didn't *start* there.

By comparison, I now rarely bother consulting MDN at all, because too much of the useful information has been removed—and so you have one less non-web-developer user, and so even less reason to maintain whatever little useful-but-not-to-web-devs information is left (and, not coincidentally, also one less pair of spec-savvy eyes to spot incorrect or misleading information such as the polyfill at issue).

(I realise this bug report probably isn't the best place for this rant, but I don't really know where would be.)
